yt/frontends/halo_catalog/data_structures.py

```diff
@@ -146,3 +177,363 @@ def _is_valid(self, *args, **kwargs):
+            ]
+        )
+
+    def _create_halo_id_table(self):
+        pass
+
+    def _detect_output_fields(self):
+        field_list = []
+        scalar_field_list = []
+        units = {}
+        pc = dict(
+            [
+                (ptype, sum([d.total_particles[ptype] for d in self.data_files]))
+                for ptype in self.ds.particle_types_raw
+            ]
+        )
+        found_fields = dict([(ptype, False) for ptype, pnum in pc.items() if pnum > 0])
+        has_ids = False
+
+        for data_file in self.data_files:
+            fl, sl, idl, _units = self.io._identify_fields(data_file)
+            units.update(_units)
+            field_list.extend([f for f in fl if f not in field_list])
+            scalar_field_list.extend([f for f in sl if f not in scalar_field_list])
+            for ptype in found_fields:
+                found_fields[ptype] |= data_file.total_particles[ptype]
+            has_ids |= len(idl) > 0
+            if all(found_fields.values()) and has_ids:
+                break
+
+        self.field_list = field_list
+        self.scalar_field_list = scalar_field_list
+        ds = self.dataset
+        ds.scalar_field_list = scalar_field_list
+        ds.particle_types = tuple(set(pt for pt, ds in field_list))
+        ds.field_units.update(units)
+        ds.particle_types_raw = ds.particle_types
+
+    def _get_halo_file_indices(self, ptype, identifiers):
+        return np.digitize(identifiers, self._halo_index_start[ptype], right=False) - 1
+
+    def _get_halo_scalar_index(self, ptype, identifier):
+        i_scalar = self._get_halo_file_indices(ptype, [identifier])[0]
+        scalar_index = identifier - self._halo_index_start[ptype][i_scalar]
+        return scalar_index
+
+    def _get_halo_values(self, ptype, identifiers, fields, f=None):
+        """
+        Get field values for halo data containers.
+        """
+
+        # if a file is already open, don't open it again
+        filename = None if f is None else f.filename
+
+        data = defaultdict(lambda: np.empty(identifiers.size))
+        i_scalars = self._get_halo_file_indices(ptype, identifiers)
+        for i_scalar in np.unique(i_scalars):
+            target = i_scalars == i_scalar
+            scalar_indices = identifiers - self._halo_index_start[ptype][i_scalar]
+
+            # only open file if it's not already open
+            my_f = (
+                f
+                if self.data_files[i_scalar].filename == filename
+                else h5py.File(self.data_files[i_scalar].filename, "r")
+            )
+
+            for field in fields:
+                data[field][target] = self._read_halo_particle_field(
+                    my_f, ptype, field, scalar_indices[target]
+                )
+
+            if self.data_files[i_scalar].filename != filename:
+                my_f.close()
+
+        return data
+
+    def _identify_base_chunk(self, dobj):
+        pass
```

Comment:
```suggestion
        raise NotImplementedError
```

Reply:
No, this will be called by `YTSelectionContainer.get_data` but is unnecessary here, so I want to pass.